source/common/quic/envoy_tls_server_handshaker.cc

```diff
@@ -0,0 +1,115 @@
+#include "source/common/quic/envoy_tls_server_handshaker.h"
+
+#include "source/common/quic/envoy_quic_proof_source.h"
+
+namespace Envoy {
+namespace Quic {
+
+EnvoyTlsServerHandshaker::EnvoyTlsServerHandshaker(
+    quic::QuicSession* session, const quic::QuicCryptoServerConfig* crypto_config,
+    Envoy::Event::Dispatcher& dispatcher)
+    : quic::TlsServerHandshaker(session, crypto_config), dispatcher_(dispatcher),
+      proof_source_(crypto_config->proof_source()) {}
+
+std::unique_ptr<quic::ProofSourceHandle> EnvoyTlsServerHandshaker::MaybeCreateProofSourceHandle() {
+  return std::make_unique<EnvoyDefaultProofSourceHandle>(
+    this, proof_source_, quic::TlsServerHandshaker::MaybeCreateProofSourceHandle());
+}
+
+EnvoyTlsServerHandshaker::EnvoyDefaultProofSourceHandle::EnvoyDefaultProofSourceHandle(
+  EnvoyTlsServerHandshaker* tls_server_handshaker,
+  quic::ProofSource* proof_source,
+  std::unique_ptr<quic::ProofSourceHandle> default_proof_source_handle) :
+    handshaker_(tls_server_handshaker),
+    proof_source_(dynamic_cast<EnvoyQuicProofSource*>(proof_source)),
+    default_proof_source_handle_(std::move(default_proof_source_handle)) {
+  ASSERT(handshaker_ != nullptr);
+}
+
+EnvoyTlsServerHandshaker::EnvoyDefaultProofSourceHandle::~EnvoyDefaultProofSourceHandle() {
+  CloseHandle();
+}
+
+// Close the handle. Cancel the pending signature operation, if any.
+void EnvoyTlsServerHandshaker::EnvoyDefaultProofSourceHandle::CloseHandle() {
+  if (private_key_method_provider_ == nullptr) {
+    default_proof_source_handle_->CloseHandle();
+  }
+}
+
+// Delegates to proof_source_->GetCertChain.
+// Returns QUIC_SUCCESS or QUIC_FAILURE. Never returns QUIC_PENDING.
+quic::QuicAsyncStatus EnvoyTlsServerHandshaker::EnvoyDefaultProofSourceHandle::SelectCertificate(
+    const quic::QuicSocketAddress& server_address,
+    const quic::QuicSocketAddress& client_address,
+    const quic::QuicConnectionId& original_connection_id, absl::string_view ssl_capabilities,
+    const std::string& hostname, absl::string_view client_hello, const std::string& alpn,
+    std::optional<std::string> alps, const std::vector<uint8_t>& quic_transport_params,
+    const std::optional<std::vector<uint8_t>>& early_data_context,
+    const quic::QuicSSLConfig& ssl_config) {
+  return default_proof_source_handle_->SelectCertificate(
```

Comment:
I think default_proof_source_handle_ is not needed if you pass in transport_socket_factory_ which knows about how to select certs.

Reply:
yea, we can just copy the code from here:
https://github.com/google/quiche/blob/main/quiche/quic/core/tls_server_handshaker.cc#L91-L125

I'm hesitate whether we should re-implement it. since we already have default proof source handle here. why don't reuse it.

Reply:
You won't need to duplicate the QUICHE code as that code also calls into Envoy proof source to find a transport socket factory and call `getTlsCertificateAndKey()` on it. If you pass down transport socket factory already, you can call `getTlsCertificateAndKey()` directly.